### ISSUE TITLE
Questions about poly encoder

### ISSUE THREAD

**Issue by min942773** [facebookresearch/ParlAI#3187]
Hi, I am using Poly-Encoder Transformer Convai2 Model and I have some questions about how it works.

1) Where does it preprocess persona-chat files? (For example, train_self_original.txt, train_self_revised.txt, ...)
2) Where does it calculate loss and eval_accuracy?

I really want to know the exact file, but I couldn't find it.
Thank you.

**Comment by klshuster** [comment]
Hi there,

1) The ConvAI2 teachers are in [this file](https://github.com/facebookresearch/ParlAI/blob/master/parlai/tasks/convai2/agents.py) (see the full directory for build details); the teachers subclass [this base teacher](https://github.com/facebookresearch/ParlAI/blob/d7361333f05d8ca108bfb5133b80b2f123e17d64/parlai/core/teachers.py#L1034) (this should help understand how to process the text files).
2) During inference/evaluation, the loss is computed in the model's [`eval_step` function](https://github.com/facebookresearch/ParlAI/blob/master/parlai/core/torch_ranker_agent.py#L482). Specifically, scores are computed via [`score_candidates`](https://github.com/facebookresearch/ParlAI/blob/master/parlai/core/torch_ranker_agent.py#L512). The poly-encoder specific code lives [here](https://github.com/facebookresearch/ParlAI/blob/d7361333f05d8ca108bfb5133b80b2f123e17d64/parlai/agents/transformer/polyencoder.py); take a look at the `score_candidates` function to see how scores are produced. The scores, along with the appropriate label indices, are passed into the `torch.nn.CrossEntropyLoss` criterion; the criterion is built [here](https://github.com/facebookresearch/ParlAI/blob/master/parlai/core/torch_ranker_agent.py#L250), and used [here](https://github.com/facebookresearch/ParlAI/blob/master/parlai/core/torch_ranker_agent.py#L527)

**Comment by min942773** [comment]
Thank you for the detailed response!
